src/storage-controller/src/lib.rs

```diff
@@ -1916,6 +1922,84 @@ where
+    /// - `off` -> `on`: If `A` gets a write timestamp higher than `B`'s boot
+    ///   timestamp, then `A` can write directly to a data shard after it's been
+    ///   registered with a txn-shard, breaking the invariant that no data shard
+    ///   is written to directly while it's registered to a transaction shard.
+    ///   To mitigate this, we should do a leadership check AFTER getting the
+    ///   write timestamp. In order for `B` to register a data shard in the txn
+    ///   shard, it must first become the leader then second get a register
+    ///   timestamp. So if `A` gets a write timestamp higher than `B`'s register
+    ///   timestamp, it will fail the leadership check.
+    /// - `on` -> `off`: If `A` tries to write to the txn-shard at a timestamp
+    ///   higher than `B`'s boot timestamp, it will fail because the shards have
+    ///   been forgotten. So everything should be ok.
+    ///
+    ///  In general, all transitions make the following steps:
+    ///   1. Get write timestamp, `ts`.
+    ///   2. Apply all transactions to all data shards up to `ts`.
+    ///   3. Register/forget all data shards. So if we crash at any point in
+    ///      these steps, for example after only applying some transactions,
+    ///      then the next Coordinator can pick up where we left off and finish
+    ///      whatever needs finishing.
+    ///
+    /// H/t jkosh44 for the above notes from the discussion in which we hashed
+    /// this all out.
+    async fn init_txns(&mut self, init_ts: T) -> Result<(), StorageError> {
+        info!(
+            "init_txns at {:?}: enable_txn={} txns={}",
+            init_ts, self.enable_persist_txn_tables, self.txns_id
+        );
+        assert_eq!(self.txns_init_run, false);
+
+        let mut txns = TxnsHandle::<SourceData, (), T, i64, PersistEpoch, TxnsCodecRow>::open(
+            T::minimum(),
+            self.txns_client.clone(),
+            self.txns_id,
+            Arc::new(RelationDesc::empty()),
+            Arc::new(UnitSchema),
+        )
+        .await;
+
+        // If successful, this forget_all call guarantees:
+        // - That we were able to write to the txns shard at `init_ts` (a
+        //   timestamp given to us by the coordinator).
+        // - That no data shards are registered at `init_ts` and thus every
+        //   table is now free to be written to directly at times greater than
+        //   that. This is not necessary if the txns feature is enabled, we
+        //   could instead commit an empty txn, but we need the apply guarantee
+        //   that it has and it doesn't hurt to start everything from a clean
+        //   slate on boot (register is idempotent and create_collections will
+        //   be called shortly).
+        // - That all txn writes through `init_ts` have been applied
+        //   (materialized physically in the data shards).
+        let removed = txns
+            .forget_all(init_ts)
```

Comment:
Something I realized after we talked is that forget_all will only (can only) advance the upper of the things that are registered at the given ts. I've already started paging this all out (and my brain is already a bit on holiday), but I can't remember if we ended up needing to ensure that the upper of every table shard is advanced past the init_ts, or whether the check_leadership in the write path means we don't

Reply:
Answer is we don't need this.